Improved Derby and CommonClassLoaderService initialization

dmatej opened the pull request:
- The Install root is resolved from the `StartupContext`, however it is not mandatory.
- `AppServerStartupTest` intentionally does not depend on `KernelJUnitExtension`, I did not notice that when I made previous changes
- `KernelJUnitExtension` should not set any System properties
- `HK2JUnit5Extension` uses own `ServerEnvironmentImpl` extension which provides the preconfigured `StartupContext` and not null.
- Improved `DriverLoader` and its landscape
- Introduced `org.glassfish.connectors.dbVendorMappingRoot` system property
- `DriverLoader` is a singleton now
- Improved resource management
- Reduced copy and paste

dmatej committed:
Improved Derby and CommonClassLoaderService initialization

- ACCStartupContext mixes asenv with JVM options
- ACCStartupContext's original code used AS_DERBY_INSTALL instead of
  com.sun.aas.derbyRoot property, which should be preferred in StartupContext.
- com.sun.aas.derbyRoot was moved from SystemPropertyConstants to BootstrapKeys
  - It is not so easy to move these constants, but we should clean this mess
  - Original constant made deprecated, is used just in ASenvPropertyReader

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

OndroMih commented on nucleus/core/kernel/src/main/java/com/sun/enterprise/v3/server/CommonClassLoaderServiceImpl.java:
```diff
@@ -189,87 +178,57 @@ private static String urlsToClassPath(Stream<URL> urls) {
 
-        String derbyHome = null;
-
-        if (arguments != null) {
```
Is it OK now to ignore any `DERBY_HOME_PROP` property in `StartupContext`? Is it not used anywhere?

dmatej replied:
Seems it is set just in `ACCStartupContext`. So you are probably right, I should use better strategy for these changes and always check following changes in my branch which probably fixed that. On the other hand no test failed, GF is quite overengineered here. Or underengineered with all those dependencies crossing paths ...

dmatej replied:
I started understanding why this ... all those values have single source, however BootstrapKeys are not available on classpath everywhere. 
But still, convention of the StartupContext is to use "JVM options like" keys, not ENV like keys.

OndroMih commented on nucleus/core/kernel/src/main/java/com/sun/enterprise/v3/server/CommonClassLoaderServiceImpl.java:
```diff
@@ -189,87 +178,57 @@ private static String urlsToClassPath(Stream<URL> urls) {
-                return (name.endsWith(".jar") && !name.startsWith("derbyLocale_"));
-            }
-        }));
+        return new File(installDir, "config").toPath().resolve(derbyHome);
```
Why does this use "installDir/config" as the root path? Is it an arbitrary path introduced here if the property specifies a relative path? Should we also use it here: https://github.com/eclipse-ee4j/glassfish/blob/7bb61e595e5f43b02406a2264a7d351f2389db0e/appserver/connectors/connectors-runtime/src/main/java/com/sun/enterprise/connectors/util/DriverLoader.java#L495 ?

dmatej replied:
Aha, yes, I added that because when it is a relative path, it is always resolved to config dir. However all these relative paths we use now are bit insecure, so this is rather temporary. And maybe not needed any more, after recent changes, I have to check.

OndroMih commented on nucleus/core/kernel/src/main/java/com/sun/enterprise/v3/server/CommonClassLoaderServiceImpl.java:
```diff
@@ -189,87 +178,57 @@ private static String urlsToClassPath(Stream<URL> urls) {
-                System.getProperty(DERBY_HOME_PROP));
-        }
+        return Arrays
+            .asList(derbyLib.listFiles((dir, name) -> name.endsWith(".jar") && !name.startsWith("derbyLocale_")));
```
Is it worth to introduce a constant for `.jar`? It's also used in `JarFileFilter` class

dmatej replied:
Not sure, because we are messing with classloaders. Any new/removed dependency can change order of initialization which can break something.

dmatej committed:
Code cleanup

- Individual steps moved to methods

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej commented on appserver/connectors/connectors-runtime/src/main/java/com/sun/enterprise/connectors/util/DriverLoader.java:
```diff
@@ -492,7 +494,7 @@ private boolean isVendorSpecific(File f, String dbVendor, String className,
     List<File> jarFileLocations = new ArrayList<>();
-        jarFileLocations.add(getLocation(SystemPropertyConstants.DERBY_ROOT_PROPERTY));
+        jarFileLocations.add(getLocation(DERBY_ROOT_PROP_NAME));
         jarFileLocations.add(getLocation(SystemPropertyConstants.INSTALL_ROOT_PROPERTY));
```
Outside of my intentions in this PR. These changes cause more issues then one would expect. But I have some changes already in my stash, the problem is that this is a basic commit, so every change I do here will cause conflicts in rebase.

dmatej commented on nucleus/core/kernel/src/main/java/com/sun/enterprise/v3/server/CommonClassLoaderServiceImpl.java:
```diff
@@ -79,78 +84,62 @@ public class CommonClassLoaderServiceImpl implements PostConstruct {
-                "createCommonClassLoader",
-                KernelLoggerInfo.systemPropertyNull,
-                SystemPropertyConstants.INSTALL_ROOT_PROPERTY);
+        final String installRoot = System.getProperty(INSTALL_ROOT_PROP_NAME);
```
Actually it is a good idea, I will try that.

dmatej replied:
StartupContextUtil is deleted in 710, all these things around are much less complicated.

dmatej committed:
Improved DriverLoader and its landscape

- Introduced org.glassfish.connectors.dbVendorMappingRoot system property
- DriverLoader is a singleton now
- Improved resource management
- Reduced copy and paste

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
Removed default derby path - always resolve startup context or system property

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej committed:
CommonClassLoaderServiceImpl - fixed tests and released requirements

- The Install root is resolved from the StartupContext, however it is not
  mandatory.
- AppServerStartupTest intentionally does not depend on KernelJUnitExtension,
  I did not notice that when I made previous changes
- KernelJUnitExtension should not set any System properties
- HK2JUnit5Extension uses own ServerEnvironmentImpl extension which provides
  the preconfigured StartupContext and not null.

Signed-off-by: David Matějček <david.matejcek@omnifish.ee>

dmatej commented:
The SSHClusterITest failure - I need to rebase to the last commit of #25470 fixing that.
